File: prototype/rails/src/test/domain/cart/cart_totals_test.rb
require "test_helper"

module Domain
  module Cart
    class CartTotalsTest < ActiveSupport::TestCase
      test "it counts every item" do
        assert_equal 3, CartTotals.from([line(1, 4500, 2), line(1, 1000, 1)]).item_count
      end

      test "it adds every line" do
        assert_equal 10_000, CartTotals.from([line(1, 4500, 2), line(2, 1000, 1)]).subtotal.cents
      end

      test "it splits the subtotal by seller" do
        totals = CartTotals.from([line(2, 1000, 1), line(1, 4500, 2), line(1, 500, 1)])

        assert_equal({ 1 => 9500, 2 => 1000 }, totals.subtotals_by_seller.transform_values(&:cents))
      end

      test "it orders the sellers by id" do
        totals = CartTotals.from([line(9, 1000, 1), line(3, 1000, 1)])

        assert_equal [3, 9], totals.subtotals_by_seller.keys
      end

      test "an empty cart totals nothing" do
        totals = CartTotals.from([])

        assert_equal 0, totals.item_count
        assert_equal 0, totals.subtotal.cents
        assert_empty totals.subtotals_by_seller
      end

      test "checkout refuses an empty cart" do
        assert_raises(ArgumentError) { CartTotals.for_checkout([]) }
      end

      test "checkout totals a cart that has something in it" do
        assert_equal 4500, CartTotals.for_checkout([line(1, 4500, 1)]).subtotal.cents
      end

      private

      def line(seller_id, unit_price_cents, quantity)
        CartLine.new(seller_id: seller_id, unit_price: Money.from_cents(unit_price_cents), quantity: quantity)
      end
    end
  end
end

File: prototype/node/src/app/core/cart/cart-totals.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { fixtureId } from '../../test/fixture-ids.ts'
import { cartTotals, checkoutTotals } from './cart-totals.ts'
import type { CartLine } from './cart-line.ts'
import type { Cents } from '../money.ts'
import { cents } from '../money.ts'
import type { SellerId } from '../ids/entity-ids.ts'

function line(sellerId: SellerId, unitPriceCents: Cents, quantity: number): CartLine {
  return { sellerId, unitPriceCents, quantity }
}

test('it counts every item', () => {
  assert.equal(cartTotals([line(fixtureId('sel', 1), cents(4500), 2), line(fixtureId('sel', 1), cents(1000), 1)]).itemCount, 3)
})

test('it adds every line', () => {
  assert.equal(cartTotals([line(fixtureId('sel', 1), cents(4500), 2), line(fixtureId('sel', 2), cents(1000), 1)]).subtotalCents, 10_000)
})

test('it splits the subtotal by seller', () => {
  const totals = cartTotals([line(fixtureId('sel', 2), cents(1000), 1), line(fixtureId('sel', 1), cents(4500), 2), line(fixtureId('sel', 1), cents(500), 1)])

  assert.deepEqual(totals.subtotalsBySeller, [
    { sellerId: fixtureId('sel', 1), subtotalCents: 9500 },
    { sellerId: fixtureId('sel', 2), subtotalCents: 1000 },
  ])
})

test('it orders the sellers by id', () => {
  const totals = cartTotals([line(fixtureId('sel', 9), cents(1000), 1), line(fixtureId('sel', 3), cents(1000), 1)])

  assert.deepEqual(
    totals.subtotalsBySeller.map((subtotal) => subtotal.sellerId),
    [fixtureId('sel', 3), fixtureId('sel', 9)],
  )
})

test('an empty cart totals nothing', () => {
  const totals = cartTotals([])

  assert.equal(totals.itemCount, 0)
  assert.equal(totals.subtotalCents, 0)
  assert.deepEqual(totals.subtotalsBySeller, [])
})

test('checkout refuses an empty cart', () => {
  assert.throws(() => checkoutTotals([]), RangeError)
})

test('checkout totals a cart that has something in it', () => {
  assert.equal(checkoutTotals([line(fixtureId('sel', 1), cents(4500), 1)]).subtotalCents, 4500)
})
